src/sciline/container.py

```diff
@@ -0,0 +1,94 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+import typing
+from functools import wraps
+from typing import Callable, List, Type, TypeVar, Union
+
+import injector
+from dask.delayed import Delayed
+
+T = TypeVar('T')
+
+
+class UnsatisfiedRequirement(Exception):
+    pass
+
+
+class Container:
+    def __init__(self, inj: injector.Injector, /, *, lazy: bool) -> None:
+        self._injector = inj
+        self._lazy = lazy
+
+    def get(self, tp: Type[T], /) -> Union[T, Delayed]:
+        try:
+            # We are slightly abusing Python's type system here, by using the
+            # injector to get T, but actually it returns a Delayed that can
+            # compute T. self._injector does not know this due to how we setup the
+            # bindings. We'd like to use Delayed[T], but that is not supported yet:
+            # https://github.com/dask/dask/pull/9256
+            task: Delayed = self._injector.get(tp)  # type: ignore
+        except injector.UnsatisfiedRequirement as e:
+            raise UnsatisfiedRequirement(e) from e
+        return task if self._lazy else task.compute()
+
+
+def _delayed(func: Callable) -> Callable:
+    """
+    Decorator to make a function return a delayed object.
+
+    In contrast to dask.delayed, this uses functools.wraps, to preserve the
+    type hints, which is a prerequisite for injector to work.
+    """
+    import dask
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        return dask.delayed(func)(*args, **kwargs)
+
+    return wrapper
+
+
+def _injectable(func: Callable) -> Callable:
+    """
+    Wrap a regular function so it can be registered in an injector and have its
+    parameters injected.
+    """
+    # When building a workflow, there are two common problems:
+    #
+    # 1. Intermediate results are used more than once.
+    # 2. Intermediate results are large, so we generally do not want to keep them
+    #    in memory longer than necessary.
+    #
+    # To address these problems, we can internally build a graph of tasks, instead of
+    # directly creating dependencies between functions. Currently we use Dask for this.
+    # The Container instance will automatically compute the task, unless it is marked
+    # as lazy. We therefore use singleton-scope (to ensure Dask will recognize the
+    # task as the same object) and also wrap the function in dask.delayed.
```

Comment:
This comment should rather be part of a design document instead of hidden away in the source code.

Reply:
I can add it.